[FLOSS T3] docs: Improve Git Documentation (#4764)
ElektraInitiative/libelektra: doc/GIT.md

> # GIT

Reviewer: ```suggestion
# Git
```
see #4679

> ## Cheat Sheet: Basic GIT Commands

Reviewer: ```suggestion
## Cheat Sheet: Basic Git Commands
```

```sh
git add readme.md   # adds the changes of the file `readme.md` to the staging area
git add .           # adds all changes of files in the current directory (recursively) to the staging area
git add --all       # adds all changes of files in the repository to the staging area
git commit -a       # executes a commit that automatically stages all changed and deleted files before
git checkout -b     # Create a new local branch based on the current branch
git pull            # Pull changes from the current remote branch
git push            # Push changes to the current remote branch
```

## Installation

- [Windows](https://git-scm.com/download/win)
- [macOS](https://git-scm.com/download/mac)
- [Linux](https://git-scm.com/download/linux)

> If you are uncomfortable, with command line interfaces, there are also [plenty of GUI options available as well](https://git-scm.com/downloads/guis)

Reviewer: ```suggestion
If you are uncomfortable with command line interfaces, there are also [plenty of GUI options available](https://git-scm.com/downloads/guis).
```

> git config merge.ff false

Reviewer 2: Good improvement to change that to the repository only as it totally depends on the strategy of the upstream repository. In this merging section I would also add `git config pull.rebase true` which will automatically rebase when pulling which makes it unnecessary calling it manually.

```

> This ensures that git will always create a merge commit, when you are trying to merge

Reviewer: ```suggestion
This ensures that Git will always create a merge commit, when you are trying to merge.
```

Also ensure that you have your username and e-mail configured, so your work can be attributed to you:

```sh
git config --global user.name "Your Name"
git config --global user.email "yourname@yourdomain.com"
```

### Adding/creating an SSH Key

> GitHub supports both HTTP and SSH for git operations. Using SSH, you can remove the annoyance of having to enter your password every time. To learn how to add an SSH-Key follow [this Guide provided by GitHub](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account)

Reviewer: ```suggestion
GitHub supports both HTTP and SSH for Git operations. Using SSH, you can remove the annoyance of having to enter your password every time. To learn how to add an SSH-Key follow [this guide provided by GitHub](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account).
```

> **Note**: These commands work with any branch checked out.
> You don't need to switch to the `master` branch first.
> However, this only works, if you have not modified your `master` branch, i.e. the latest commit in your forked `master` branch was once the latest commit in the main `master` branch.

## Rebasing branches

When you work on a local feature branch, it might happen that a change occurs within the master branch, that introduces merge conflicts. I.e. if you create a pull request, it cannot be merged into master automatically anymore. In these cases, we expect you to rebase your feature branch, so it can be merged automatically once again.

To achieve this, make sure you have your master branch synced up. Then switch to the feature branch you want to update using `git checkout <feature-branch-name>` and then type `git rebase master`. Sometimes, this will result in a merge conflict.

> If you already have pushed changes from your feature branch to a remote branch, you'll need to either overwrite it using `git push --force` or push to a new remote branch

Reviewer: ```suggestion
If you already have pushed changes from your feature branch to a remote branch, you'll need to either overwrite it using `git push --force` or push to a new remote branch.
```

### Resolving merge conflicts

> To resolve these edit the files that need to be merged manually. You can check which files need to be merged using `git status`. After you are done merging a file manually, you can use `git add <path-to-file>` and when you have merged all files, continue with `git rebase --continue`.

Reviewer: ```suggestion
To resolve merge conflicts, edit the files that need to be merged manually. You can check which files need to be merged using `git status`. After you are done merging a file manually, you can use `git add <path-to-file>` and when you have merged all files, continue with `git rebase --continue`.
```

> > **Note**: Keep in mind that doing manual merges within a rebase are destructive. If you accidentally remove changes you've done in your feature branch, they will be gone forever. If you're new to manually merging files in git, it's a good idea to create a new branch from your feature branch using `git checkout -b <new-branch-name>`. When you're stuck on a manual merge, you can also always abort the rebase using `git rebase --abort`

Reviewer 3: This is not entirely true. By default Git will not remove previous commits immediately, even if they are no longer reachable from any of the branches.

Unreachable commits are only deleted when the garbage collector is invoked. (The documentation is not very specific about when this occurs.)

In any case there is `git-reflog` which allows one to retrieve previous commits even if they are no longer reachable from any branch; provided they have not been deleted yet by the garbage collector.

* <https://git-scm.com/docs/git-reflog/2.39.0>
* <https://git-scm.com/docs/git-gc>

(Basically anything you have committed is fine to overwrite using Git commands. The only dangerous thing is doing `git reset` when having changes in the working tree/index which have not been committed at any point in time.)

Reviewer 4: I agree the sentence should be corrected.

> The only dangerous thing 

As you said yourself, you lose unreachable commits after `git gc` and reflogExpireUnreachable is reached (or time jumped). So it is also not completely without any danger.

Reviewer 2: I agree on that, but I would consider them "deleted" although as the gc can delete them unexpectedly anytime. I think you can compare the situation to java where the gc will delete objects without a reference but they remain before the gc jumps in. I think almost nobody would call them not-deleted when describing this behavior to a new user who has to understand the basics before thinking about the gc.

Author: I'll try writing an explanation that's more true to life while not being too complicated

> > **Note**: Keep in mind that doing manual merges within a rebase are destructive. If you accidentally remove changes you've done in your feature branch, they will be gone forever. If you're new to manually merging files in git, it's a good idea to create a new branch from your feature branch using `git checkout -b <new-branch-name>`. When you're stuck on a manual merge, you can also always abort the rebase using `git rebase --abort`

Reviewer: ```suggestion
> **Note**: Keep in mind that manual merges within a rebase are destructive. If you accidentally remove changes you've done in your feature branch, they will be gone forever. If you're new to manually merging files in Git, it's a good idea to create a new branch from your feature branch using `git checkout -b <new-branch-name>`. When you're stuck on a manual merge, you can also always abort the rebase using `git rebase --abort`
```
